app/naive_bayes.py

```diff
@@ -0,0 +1,286 @@
+                    dict_link_features["Agent_Privilege"] = privilege
+                    dict_link_features["Host_Architecture"] = architecture
+                else:
+                    dict_link_features["Agent_Protocol"] = None
+                    dict_link_features["Trusted_Status"] = None
+                    dict_link_features["Agent_Privilege"] = None
+                    dict_link_features["Host_Architecture"] = None
+
+                dict_link_features["Link_Facts"] = self._get_useful_facts(cur_link)
+
+                link_row = [
+                    dict_link_features["Status"],
+                    dict_link_features["Ability_ID"], 
+                    dict_link_features["Link_Facts"], 
+                    dict_link_features["Planner"],
+                    dict_link_features["Obfuscator"],
+                    dict_link_features["Adversary_ID"],
+                    dict_link_features["Adversary_Name"],
+                    dict_link_features["Command"],
+                    dict_link_features["Number_Facts"],
+                    dict_link_features["Visibility_Score"],
+                    dict_link_features["Executor_Platform"],
+                    dict_link_features["Executor_Name"],
+                    dict_link_features["Agent_Protocol"],
+                    dict_link_features["Trusted_Status"],
+                    dict_link_features["Agent_Privilege"],
+                    dict_link_features["Host_Architecture"]
+                ]
+                matrix_link_data.append(link_row)
+        return matrix_link_data
+
+    def _get_useful_facts(self, cur_link : Link) -> Dict[str, str]:
+        """
+        Selects generalizable facts from link object for data querying
+        """
+        useful_facts = {}
+        if(len(cur_link.used) > 0):
+            for used_fact in cur_link.used:
+                # exclude fact types that are unique to a specific host
+                if str(used_fact.trait).startswith(("host.", "remote.", "file.last.", "domain.user.")) == False:
+                    useful_facts[str(used_fact.trait)] = str(used_fact.value)
+        return useful_facts
+
+    def _query_link_matrix(self, matrix_link_data : List[List], query_features : Dict) -> List[List]:
+        """
+        Query link matrix for links containing specific features and values and 
+        return new matrix composed of column names row and the link rows containing right feat and val
+
+        :param matrix_link_data: link matrix that will be queried
+        :param query_features: dict of {feature_names: values} that matrix_link_data will be queried by
+        """
+        col_name_to_col_index = {}
+        list_col_names = matrix_link_data[0]
+        for index_feature_names in range(len(list_col_names)):
+            col_name_to_col_index[list_col_names[index_feature_names]] = index_feature_names
+
+        queried_link_matrix = [list_col_names]
+        for row_index in range(1, len(matrix_link_data)):
```

Comment:
Related to the last two comments, if you don't have feature names as the first element of the list, you wouldn't need to do this `range(1, len(matrix_link_data))` thing. You could just iterate the list like normal since it would all be of the same type.